lib/auth/oidc/provider.go

```diff
@@ -0,0 +1,189 @@
+package oidc
+
+import (
+	"context"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/hashicorp/cap/oidc"
+	"github.com/hashicorp/nomad/nomad/structs"
+)
+
+// providerConfig returns the OIDC provider configuration for an OIDC
+// auth-method.
+func providerConfig(authMethod *structs.ACLAuthMethod) (*oidc.Config, error) {
+	var algs []oidc.Alg
+	if len(authMethod.Config.SigningAlgs) > 0 {
+		for _, alg := range authMethod.Config.SigningAlgs {
+			algs = append(algs, oidc.Alg(alg))
+		}
+	} else {
+		algs = []oidc.Alg{oidc.RS256}
+	}
+
+	return oidc.NewConfig(
+		authMethod.Config.OIDCDiscoveryURL,
+		authMethod.Config.OIDCClientID,
+		oidc.ClientSecret(authMethod.Config.OIDCClientSecret),
+		algs,
+		authMethod.Config.AllowedRedirectURIs,
+		oidc.WithAudiences(authMethod.Config.BoundAudiences...),
+		oidc.WithProviderCA(strings.Join(authMethod.Config.DiscoveryCaPem, "\n")),
+	)
+}
+
+// ProviderCache is a cache for OIDC providers. OIDC providers are something
+// you don't want to recreate per-request since they make HTTP requests
+// themselves.
+//
+// The ProviderCache purges a provider under two scenarios: (1) the
+// provider config is updated, and it is different and (2) after a set
+// amount of time (see cacheExpiry for value) in case the remote provider
+// configuration changed.
+type ProviderCache struct {
+	providers map[string]*oidc.Provider
+	mu        sync.RWMutex
+
+	// cancel is used to trigger cancellation of any routines when the cache
+	// has been informed its parent process is exiting.
+	cancel context.CancelFunc
+}
+
+// NewProviderCache should be used to initialize a provider cache. This
+// will start up background resources to manage the cache.
+func NewProviderCache() *ProviderCache {
+
+	// Create a context, so a server that is shutting down can correctly
+	// shut down the cache loop and OIDC provider background processes.
+	ctx, cancel := context.WithCancel(context.Background())
+
+	result := &ProviderCache{
+		providers: map[string]*oidc.Provider{},
+		cancel:    cancel,
+	}
+
+	// Start the cleanup timer
+	go result.runCleanupLoop(ctx)
+
+	return result
+}
+
+// Get returns the OIDC provider for the given auth method configuration.
+// This will initialize the provider if it isn't already in the cache or
+// if the configuration changed.
+func (c *ProviderCache) Get(authMethod *structs.ACLAuthMethod) (*oidc.Provider, error) {
+
+	// No matter what we'll use the config of the arg method since we'll
+	// use it to compare to existing (if exists) or initialize a new provider.
+	oidcCfg, err := providerConfig(authMethod)
+	if err != nil {
+		return nil, err
+	}
+
+	// Get any current provider for the named auth-method.
+	var (
+		current *oidc.Provider
+		ok      bool
+	)
+
+	// This may be the first time the cache is being used, therefore we need to
+	// ensure we don't panic when attempting to access the map.
```

Comment:
Isn't this guaranteed by the constructor?

Reply:
Resolved within commit https://github.com/hashicorp/nomad/pull/15764/commits/ebc76d23da5f673adfdfb993d0b537376d9d7e4a